### ISSUE TITLE
Teeing into two different merges

### ISSUE THREAD

**Issue by johnsusek** [pull-stream/pull-stream#124]
Hello, first of all, thank you for pull-streams, I've been trying a lot of stream libraries lately and yours has the most clarity. 

I am experimenting with turning a DAG (Directed Acyclic Graph) of instructions into a pull-stream pipeline. The graph is stored in JSON as nodes and connections. I think this is interesting because these methods could be built upon for a general purpose way to describe workflow-style computations declaratively. A node can be a merge, tee, or just an instruction/step. Right now I walk the tree recursively and build all the branches as pipelines, then I create all the tees, finally I create all the merges. 

I'm explaining this in case you're wondering about this weird use case I'm describing. This method works fairly well, but there is an issue I'm having in the case of a tee feeding into two different merges.

The examples below just use a stream of json objects from thecatapi.com. I'll add diagrams below for clarity, I hope they help.

First, I wanted to show how I am using tees and merges together. Given a structure like this:
![Screen Shot 2019-12-10 at 11 56 43 PM](https://user-images.githubusercontent.com/611996/70595325-bac2b680-1ba8-11ea-84d3-2cacf3e7bbe8.png)

Here is an example of building that:

```
// The "M" shape
// Combined fork and merge
(async () => {
  let jsonStreamA = await fetchJson({
    selector: '*',
    url: 'https://api.thecatapi.com/v1/breeds'
  });

  let jsonStreamB = await fetchJson({
    selector: '*',
    url: 'https://api.thecatapi.com/v1/images'
  });

  let Fork1ToCount1 = pull(pull.map(count()), pull.collect(console.log));
  let Fork2ToCount2 = pull(pull.map(count()), pull.collect(console.log));
  let S1toFork1 = pull(jsonStreamA, tee(Fork1ToCount1));
  let S2toFork2 = pull(jsonStreamB, tee(Fork2ToCount2));
  let MergedForks = many([S1toFork1, S2toFork2]);
  pull(MergedForks, pull.map(count()), pull.collect(console.log));
})();
```

 So that works and demonstrates mixing tees and merges together. 

So, finally on to the issue I am having! There could be a scenario where a tee can feed into two different merges, like if we add these gray boxes to the above diagram:

![Screen Shot 2019-12-11 at 12 40 20 AM](https://user-images.githubusercontent.com/611996/70597774-d466fc80-1bae-11ea-8feb-778fda444b33.png)

Now the right-hand tee is feeding into two different merges. 

I've been trying something along these lines with no success:

```
  let jsonStreamA = await fetchJson({
    selector: '*',
    url: 'https://api.thecatapi.com/v1/breeds'
  });

  let jsonStreamB = await fetchJson({
    selector: '*',
    url: 'https://api.thecatapi.com/v1/images'
  });

  let jsonStreamC = await fetchJson({
    selector: '*',
    url: 'https://api.thecatapi.com/v1/categories'
  });

  let Fork1ToCount1 = pull(pull.map(count()), pull.collect(console.log));
  let Fork2ToCount2 = pull(pull.map(count()), pull.collect(console.log));
  let S1toFork1 = pull(jsonStreamA, tee(Fork1ToCount1));
  let foo = tee(); // added
  let S2toFork2 = pull(jsonStreamB, tee(Fork2ToCount2), foo); // changed
  let MergedForks = many([S1toFork1, S2toFork2]);
  pull(MergedForks, pull.map(count()), pull.collect(console.log));
  let NewMerge = many([jsonStreamC, foo]); // the new gray box merge
  pull(NewMerge, pull.map(count()), pull.collect(console.log)); // the new gray box count
```

That didn't work though, I've tried various values for foo, things like `tee(pull())` etc. I'm just not sure if what I am trying is even possible, which is why I'm reaching out here.

TL;DR can you think a way of making a tee split off into two merges? Is is possible? Thanks for reading all this!

**Comment by dominictarr** [comment]
oh you used `foo` as a sink twice, that doesn't work.

I think this would be significantly simplified if you made count into a through stream, then you could mostly get rid of the `tee`'s if not completely get rid of them.

btw, by using await on the `fetchJson` calls, you are actually doing A, then B, then C, and it's not streaming at all, it takes 3x the latency instead of pipelining down a single tcp connection.

**Comment by dominictarr** [comment]
make count something more like:
```
var c = 0
pull.through(function (v) { c++ })
```
instead of using `pull.map`

**Comment by johnsusek** [comment]
Here is my count implementation right now, 

```
function count() {
  let c = 0;

  return (data, cb) => {
    c += 1;
    if (cb) {
      cb(null, c);
    }
    return c;
  };
}
```

I'll change this into a through stream like you mentioned. Conceptually, I guess I was thinking of the counts as sinks, which sounds like it is not correct.


And as a side note, here is my fetchJson which you can see only awaits for the stream to start, not the entire response. Maybe this isn't the best way to stream json objects though.

```
async function fetchJson({ url, headers, selector }) {
  let res = await axios({
    method: 'get',
    url,
    responseType: 'stream',
    headers
  });

  return toPull.source(res.data.pipe(JSONStream.parse(selector)));
}
```

**Comment by johnsusek** [comment]
So I understand conceptually, the count would become a through stream, so it wouldn't be teeing, it would just collect the count as records from S1 pass through it.

Visually, it would change to something like this? Circles represent through streams 

![Screen Shot 2019-12-11 at 6 34 02 PM](https://user-images.githubusercontent.com/611996/70672408-cf548c80-1c44-11ea-98fb-c8d4798843fd.png)

No more tees, but I still can get the count of S1 and S2. So this does collect the counts for the scenario I described, and is indeed simpler.

Unfortunately, my goal is a system where users could arbitrarily fork or merge streams from any step, if they want to modify their workflow. (I am not sure if this is feasible, more of a research/experimentation phase right now.) The scenario I mentioned is a case that wasn't working for me, but what I'm aiming for is a generic algorithm that can build a valid pipeline given any list of steps and connections (as long as they are a DAG). 

So, to this end, let's imagine a scenario where we have a UI like my flowcharts, where users could branch off from a step with a mouse click. Let's say a user later wants to double the count, so the graph would look like this now:

![Screen Shot 2019-12-11 at 6 33 36 PM](https://user-images.githubusercontent.com/611996/70672421-d380aa00-1c44-11ea-8d4b-8fc21177a5f6.png)

Now that those counts are through streams, how could I add this doubling logic to the pipeline? 

Just to reiterate this is obviously a contrived pipeline, just trying to work out a general algorithm (if such a thing is possible).

I think it's really exciting though if this is possible, which is why I'm pursuing it! Thanks again!

**Comment by johnsusek** [comment]
@dominictarr Thanks, your suggestion to use through streams has led me down a fruitful path. I appreciate the help!

**Comment by dominictarr** [comment]
@johnsusek just happy I could help make your program simpler!
